File: common/src/main/java/xyz/chunkstories/net/Connection.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.net;

import java.io.File;
import java.io.IOException;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import xyz.chunkstories.api.exceptions.PacketProcessingException;
import xyz.chunkstories.api.exceptions.net.IllegalPacketException;
import xyz.chunkstories.api.net.Packet;
import xyz.chunkstories.api.net.packets.PacketText;

public abstract class Connection {
	protected final String remoteAddress;
	protected final int port;

	protected static final Logger logger = LoggerFactory.getLogger("net.connection");

	public Connection(String remoteAddress, int port) {
		super();
		this.remoteAddress = remoteAddress;
		this.port = port;
	}

	public String getRemoteAddress() {
		return remoteAddress;
	}

	public int getPort() {
		return port;
	}

	public abstract boolean connect();

	public abstract void flush();

	public abstract void handleDatagram(LogicalPacketDatagram datagram)
			throws IOException, PacketProcessingException, IllegalPacketException;

	public abstract boolean handleSystemRequest(String msg);

	public void sendTextMessage(String string) {
		PacketText packet = new PacketText();
		packet.text = string;
		pushPacket(packet);
	}

	public abstract void pushPacket(Packet packet);

	public abstract PacketsEncoderDecoder getEncoderDecoder();

	public abstract boolean close();

	public abstract boolean isOpen();

	// Below is stuff for downloading/uploading of files

	Map<String, PendingDownload> fileStreamingRequests = new ConcurrentHashMap<>();

	/**
	 * Hints the connection logic that a file with a certain tag is to be expected,
	 * and provides it with a location to save it. Unexpected file streaming will be
	 * discarded.
	 */
	public void registerExpectedFileStreaming(String fileTag, File whereToSave, DownloadStartAction action) {
		if (fileStreamingRequests.putIfAbsent(fileTag, new PendingDownload(whereToSave, action)) == null) {

		} else {
			logger.warn("Requesting twice file: " + fileTag);
		}
	}

	public void registerExpectedFileStreaming(String fileTag, File whereToSave) {
		registerExpectedFileStreaming(fileTag, whereToSave, null);
	}

	/** Retreives and removes the expected location for the specified fileTag */
	public PendingDownload getLocationForExpectedFile(String fileTag) {
		return fileStreamingRequests.remove(fileTag);
	}

	public interface DownloadStatus {

		public int bytesDownloaded();

		public int totalBytes();

		/** Waits for the download to end and returns true if successful */
		public boolean waitForEnd();
	}

	public interface DownloadStartAction {
		public void onStart(DownloadStatus status);
	}

	public class PendingDownload {
		public final File f;
		public final DownloadStartAction a;

		public PendingDownload(File f, DownloadStartAction a) {
			super();
			this.f = f;
			this.a = a;
		}
	}
}

File: common/src/main/java/xyz/chunkstories/net/Connection.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.net

import org.slf4j.LoggerFactory
import xyz.chunkstories.EngineImplemI
import xyz.chunkstories.api.Engine
import xyz.chunkstories.api.client.Client
import xyz.chunkstories.api.exceptions.PacketProcessingException
import xyz.chunkstories.api.exceptions.net.IllegalPacketException
import xyz.chunkstories.api.net.Packet
import xyz.chunkstories.api.net.packets.PacketText
import xyz.chunkstories.api.server.UserConnection
import xyz.chunkstories.api.world.WorldSub
import xyz.chunkstories.world.WorldImplementation
import xyz.chunkstories.world.WorldSubImplementation

import java.io.File
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap

typealias DownloadStartAction = (Connection.DownloadStatus) -> Unit

abstract class Connection(val engine: Engine, val remoteAddress: String, val port: Int) {

    abstract val encoderDecoder: PacketsEncoderDecoder

    abstract val isOpen: Boolean

    // Below is stuff for downloading/uploading of files
    private val fileStreamingRequests = ConcurrentHashMap<String, PendingDownload>()

    abstract fun connect(): Boolean

    abstract fun flush()

    abstract val world: WorldImplementation?
    abstract val userConnection: UserConnection?

    @Throws(IOException::class, PacketProcessingException::class, IllegalPacketException::class)
    fun handleDatagram(datagram: LogicalPacketDatagram) {
        val definition = datagram.packetDefinition as PacketDefinition
        if (definition.constructorTakesWorld) {
            val world = world
            if (world == null)
                logger.error("Received packet $definition but no world is up yet !")

            if(world is WorldSubImplementation)
                world.queueDatagram(datagram)
        } else {
            val packet = definition.createNewWithEngine(engine is Client, engine)
            packet!!.receive(datagram.data, userConnection)
            if (packet is PacketText) {
                handleSystemRequest(packet.text)
            }
            datagram.dispose()
        }
    }

    abstract fun handleSystemRequest(message: String): Boolean

    open fun sendTextMessage(string: String) {
        val packet = PacketText(engine)
        packet.text = string
        pushPacket(packet)
    }

    abstract fun pushPacket(packet: Packet)

    abstract fun close(reason: String)

    /**
     * Hints the connection logic that a file with a certain tag is to be expected,
     * and provides it with a location to save it. Unexpected file streaming will be
     * discarded.
     */
    fun registerExpectedFileStreaming(fileTag: String, whereToSave: File, action: DownloadStartAction) {
        if (fileStreamingRequests.putIfAbsent(fileTag, PendingDownload(whereToSave, action)) != null) {
            logger.warn("Requesting twice file: $fileTag")
        }
    }

    /**
     * Retreives and removes the expected location for the specified fileTag
     */
    fun getLocationForExpectedFile(fileTag: String): PendingDownload? {
        return fileStreamingRequests.remove(fileTag)
    }

    interface DownloadStatus {
        fun bytesDownloaded(): Int
        fun totalBytes(): Int
        fun waitsUntilDone(): Boolean
    }

    inner class PendingDownload(val f: File, val a: DownloadStartAction)

    companion object {
        val logger = LoggerFactory.getLogger("net.connection")
    }
}
